### ISSUE TITLE
The big 1.0 

### ISSUE THREAD

**Issue by lerno** [c3lang/c3c#1456]
I'm collecting small pieces of things needed for 1.0 that aren't really ready to go into issues yet into this single issue.

- [ ] Full BSD-family support in the standard library
- [ ] C-to-C3 tool available
- [ ] C3docs available
- [ ] Full standard library available on the web
- [ ] No known bugs in the compiler
- [ ] A secondary backend for fast compilations
- [ ] Automatic static / dynamic library .c3l generation
- [ ] A completely fleshed out standard library
- [ ] Support for most common embedded targets (when 1.0 is released!)
- [ ] Full Android / iOS support
- [ ] Downloadable binaries for all Linux distros, MacOS, Windows, all BSD variants.
- [ ] A good experience cross compiling to Linux
- [ ] Cross compiling to MacOS & Windows with automatic download of required supporting libraries

Nice to haves:
- [ ] A C output backend

**Comment by Caleb-o** [comment]
Very exciting!

**Comment by Real-Packet** [comment]
ZIP compression? *yoinks C# ZIP library implementation*

**Comment by Real-Packet** [comment]
> I'm collecting small pieces of things needed for 1.0 that aren't really ready to go into issues yet into this single issue.
> 
>     * [ ]  Full BSD-family support in the standard library
> 
>     * [ ]  C-to-C3 tool available
> 
>     * [ ]  C3docs available
> 
>     * [ ]  Full standard library available on the web
> 
>     * [ ]  No known bugs in the compiler
> 
>     * [ ]  A secondary backend for fast compilations
> 
>     * [ ]  Automatic static / dynamic library .c3l generation
> 
>     * [ ]  A completely fleshed out standard library
> 
>     * [ ]  Support for most common embedded targets (when 1.0 is released!)
> 
>     * [ ]  Full Android / iOS support
> 
>     * [ ]  Downloadable binaries for all Linux distros, MacOS, Windows, all BSD variants.
> 
>     * [ ]  A good experience cross compiling to Linux
> 
>     * [ ]  Cross compiling to MacOS & Windows with automatic download of required supporting libraries
> 
>     * [ ]  High quality LSP
> 
>     * [ ]  Easily available editor plugins
> 
>     * [ ]  High quality IntelliJ plugin
> 
> 
> Nice to haves:
> 
>     * [ ]  A C output backend
> 
> 
> Stdlib features
> 
>     * [ ]  Zip compression
> 
>     * [ ]  Regex library

Also, "Downloadable binaries for all Linux distros" is already an issue: #1436

**Comment by lerno** [comment]
> ZIP compression? _yoinks C# ZIP library implementation_

If you want to. There are a lot of them out there.

**Comment by rodgomesc** [comment]
What is still needed to complete Android support? Is there a to-do list available anywhere?

**Comment by lerno** [comment]
Something like this @rodgomesc 
1. Make sure linking links the proper libraries
2. Decide what Android platforms to support (32 bit ARM for example?)
3. Add CI for Android compilation
4. Fix whatever special ABI changes Android has compared to Linux
5. Linking should work off the bat, so there should be a way to pull in the needed libraries for linking, initially perhaps as an external script, but later as something in the compiler itself which downloads needed dependencies.
6. Add some bindings to essential Android libraries.

**Comment by radekm** [comment]
>  Zip compression

This also means ZIP decompression? If that is so, I wonder how to deal with different kinds of ZIP archives produced by different implementations. Recently I watched [Josh Wolfe's talk](https://www.youtube.com/watch?v=V-VO0Ua_b4g) about yauzl library and invalid ZIP archives produced by Mac Archive Utility (you can see the GitHub issue https://github.com/thejoshwolfe/yauzl/issues/69). It seemed that creating a reliable ZIP decompressor which works with most ZIP archives is a really hard task.

**Comment by lerno** [comment]
I think a best effort thing would be reasonable @radekm. Something on par with Java's built in zip library code.

**Comment by rodgomesc** [comment]
Hey! I started to poke around and I start thinking about something - is it possible to use CMake to turn `.c3` code into a shared library without the prebuilt c3c compiler? Ideally, I don't want users to need a c3c compiler already installed on their computer. Maybe I could go from `.c3` files to `LLVM IR`, then to `object files`, and finally into a `.so` library? Or am I just making this way more complicated than it needs to be?

**Comment by lerno** [comment]
Well, turning a .c3 file into a IR or a library is exactly what you need a compiler for. So I'm not sure how that could work @rodgomesc

**Comment by alexveden** [comment]
`c3fmt` now is released https://github.com/alexveden/c3tools/tree/main/src/c3fmt

**Comment by LukeGary462** [comment]
using C3 for embedded systems development on arm cortex mcus is extremely exciting to me. Is there an option to compile c3 without linking libc or specifying a different libc implementation? I cant wait to see how support for embedded targets progresses!

**Comment by LowByteFox** [comment]
Unsure if someone mentioned it, but for BSD binary releases, be aware that each release, the respective BSD may not be ABI compatible

let's say between OpenBSD 7.6 to 7.7

one thing that could be done is to add c3c port to their ports trees

it is also why each release, every single package is recompiled

**Comment by lerno** [comment]
You're talking about the compiler or something else?

**Comment by LowByteFox** [comment]
> You're talking about the compiler or something else?

yes, the compiler binary releases,
the standard library should be fine

**Comment by lerno** [comment]
A good point. Ideally it compiles to a wide range of targets out of the box, but also is available from different package managers.

**Comment by lerno** [comment]
What intermediate language @kryzet

**Comment by kryzet** [comment]
I mean eliminating LLVM and going directly from C3 to assembly (maybe machine code?)

**Comment by lerno** [comment]
@kryzet That's the "secondary backend" point.

**Comment by lerno** [comment]
Typically a few months. @kryzet

**Comment by kryzet** [comment]
The compiler should utilize the Sea of Nodes algorithm; that would make it competitively performant without relying on GCC/LLVM/QBE.

**Comment by LowByteFox** [comment]
> The compiler should utilize the Sea of Nodes algorithm; that would make it competitively performant without relying on GCC/LLVM/QBE.

Yes, but wouldn't say that's always the case

Look at V8
https://v8.dev/blog/leaving-the-sea-of-nodes

I know I'm comparing apples and pears but I don't think there always is one solution that's the fastest

**Comment by kryzet** [comment]
Since a number of the issues noted in that article are caused by the design of JavaScript, it is likely that this won't be an issue for C3 as it is not trying to be similar to JavaScript (though I'm not familiar with the language so I can't say for sure that there aren't any deal breakers). Do you have any other suggestions for the IR that the C3 compiler should use?

**Comment by mayurathavale18** [comment]
I am excited to work on the LSP! Let's make this a great language with backward compatibility!

**Comment by nomota** [comment]
* a decent package manager

* least dependancy on external tools like cmake, clang, llvm

* self hosted compiler completely written in C3

* complete libc replacement for removing dependancies

**Comment by VisenDev** [comment]
> * a decent package manager

I don't know what lerno's position on this is, but I generally take Ginger Bill's strong anti package manager view. I don't think a package manager is a requirement for 1.0 or even a good idea necessarily. Package managers make it way to easy to get into dependency hell, and vendoring your dependencies is usually the better choice. https://www.gingerbill.org/article/2025/09/08/package-managers-are-evil/

> * self hosted compiler completely written in C3

I don't think it is a good idea to create a self hosted compiler when the compiler in question is also the only implementation of the language. Having a self hosted compiler makes it so much more difficult to bootstrap to new targets. Some operating systems (Guix for example) and users prefer to build all their software from source. A self hosting compiler makes this otherwise trivial task almost impossible, see this blog post about how much trouble it was to reproduce the zig compiler's binaries since the zig compiler is self hosted https://jakstys.lt/2024/zig-reproduced-without-binaries/

**Comment by mtalexan** [comment]
This got longer than I expected.  
TL;DR  
The author argues to eliminate the option for the proven-successful package manager feature on the grounds it might be misused, then demonstrates this attitude comes from a deep seated sense of entitlement to bug-free open source software without contributing back and a frustration with the world not working that way.

---
 
> > * a decent package manager
> 
> I don't know what lerno's position on this is, but I generally take Ginger Bill's strong anti package manager view. I don't think a package manager is a requirement for 1.0 or even a good idea necessarily. Package managers make it way to easy to get into dependency hell, and vendoring your dependencies is usually the better choice. https://www.gingerbill.org/article/2025/09/08/package-managers-are-evil/
 
I would argue this would be an important feature actually. 

The linked article basically argues package managers are bad because they allow well modularized reuse (which introduces large dependency trees, which any well written code would have in some form no matter what), and individuals misuse public code so it bites them. Their solution is to put almost everything into a single monolithic "trustworthy" standard library, and/or intentionally add friction to package reuse so it's harder and presumably requires more consideration of when it gets used.  Both of these are classic thought traps that we know do not work out.  
I'll grant the author that there are some well known cases of dependency hell (especially in JavaScript), but that simply means you as the developer failed to do the due diligence the author says you should have. Garbage in, garbage out. Throwing out code reuse (package managers), or making it harder isn't the solution. The very case they cite of having package manager managers is exactly what that leads to when multiple someone elses decide to fix the friction. Instead the obvious solution is to make doing that due diligence easier. Solutions that warn if you only use a small part of a dependency tree or similar are the actual solution here.  

The author also complains about low quality packages being easily used. Which is interesting since the counterargument studies for using memory managed languages like Golang or Rust vs C/C++ specifically found that the leading cause of the memory management issues in C/C++ was the fragmentation that led to buggy reimplementations. Getting all eyes contributing to shared code is demonstrated to produce significantly less buggy and better performing code than fragmented reinvent-the-wheel code like the author recommends. But it requires actual open source contribution rather than entitled attitudes like the author clearly demonstrates with thier reference to the SDL2 packages being buggy and thier solution being to _roll thier own private copy instead of contributing fixes_.  

No matter how unjustifiably entitled you feel to bug-free free open source software you didn't help on/with, you can't argue that having the option for it to exist is better than not. Which is what the author is arguing against.

**Comment by gingrspacecadet** [comment]
How much of the original checklist still stands?

**Comment by gingrspacecadet** [comment]
Also, on the package manager, a well designed one can almost entirely eliminate dependency hell!

**Comment by XiaochaoYan** [comment]
Looking forward for the C-to-C3 tools. It would be very helpful for GPU programming and game dev.

**Comment by tonis2** [comment]
@kryzet "Free, open-source software does not need a package manager to exist."

But doesent C have a package manager too, it's just hidden away to `sudo apt-get` ..etc, to download C libraries, and it's still a mess sometimes.

If having a centralised host, for packages/libraries is something we don't want,
then just having something in the `c3c` that can read `project.json` and `manifest.json` gather the git urls of libraries, if provided and auto-download the libraries from git to dependencies, I think that would be a big improvement / simplification too.

**Comment by ManuLinares** [comment]
> Full standard library available on the web

Done. https://c3-lang.org/stdlib

>  High quality LSP

Getting there: https://github.com/ManuLinares/c3_ls

> C-to-C3 tool available

https://github.com/ManuLinares/c3_ls#bindgen-flags

>  A C output backend

https://github.com/c3lang/c3c/pull/3504
needs testers: `c3c compile --backend=c file.c3` || `c3c build --backend=c`
